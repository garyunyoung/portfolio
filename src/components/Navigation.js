import React, { useState, useEffect } from 'react';
import { useLocation } from 'react-router-dom'

import githubLogo from '../assets/logos/github.svg';
import codewarsLogo from '../assets/logos/codewars.svg';
import exercismLogo from '../assets/logos/exercism.svg';

import elementInView from "../utilities/elementInView";

import '../styles/Navigation.scss';

export default function Navigation() {
  const [isOpen, setIsOpen] = useState(false)
  const [isSocialsOpen, setIsSocialsOpen] = useState(false)

  const isHome = useLocation().pathname === '/'

  function toggleNav() {
    setIsOpen(prevState => !prevState)
  }

  function closeNav() {
    setIsOpen(false)
  }

  function toggleSocials() {
    setIsSocialsOpen(prevState => !prevState)
  }

  useEffect(() => {
    if (isHome) {
      elementInView()
    }
  })

  return (
    <header className={`site-header ${isOpen ? 'is-open' : ''}`}>
      <div className='site-header__content'>
        <div className='site-header__title-wrapper'>
          {isHome ?
            <h1 className='site-header__title'>Garyun Young</h1> :
            <a className={`site-header__title-link ${isHome ? '' : 'is-active'}`} href='/#'>Garyun Young</a>
          }
        </div>
        <button
          className='site-header__mobile-menu-button'
          onClick={() => toggleNav()}
        >
          {isOpen ? 'Close' : 'Menu'}
        </button>
        <NavMenu
          isSocialsOpen={isSocialsOpen}
          toggleSocials={() => toggleSocials()}
          closeNav={() => closeNav()} />
      </div>
    </header>
  );
}

function NavMenu(props) {
  return (
    <nav className='site-menu site-header__nav'>
      <ul className='site-menu__nav-items'>
        <span className="site-menu__nav-main-items">
          <li className='site-menu__nav-item'>
            <a className='site-menu__nav-item-link site-menu__nav-item-link--projects' href='/#projects' onClick={props.closeNav}>
              Projects
            </a>
          </li>
          <li className='site-menu__nav-item'>
            <a className='site-menu__nav-item-link site-menu__nav-item-link--about' href='/#about' onClick={props.closeNav}>
              About
            </a>
          </li>
        </span>
        <span className='site-menu__nav-socials-container'>
          <li
            onClick={props.toggleSocials}
          >
            <button className={`
              site-menu__nav-item-link 
              site-menu__nav-item-link--socials 
              ${props.isSocialsOpen ? 'is-selected is-active' : ''}
            `}>
              Code
            </button>
          </li>
          <span className={`site-menu__nav-socials ${props.isSocialsOpen ? 'is-visible' : ''}`}>
            <li>
              <a
                className='site-menu__nav-item-link-social'
                href='https://github.com/garyunyoung'
                target='_blank'
                rel='noopener noreferrer'
              >
                <p className="site-menu__nav-item-text-social">Github</p>
                <img src={githubLogo} alt='' />
              </a>
            </li>
            <li>
              <a
                className='site-menu__nav-item-link-social'
                href='https://exercism.org/profiles/garyunyoung'
                target='_blank'
                rel='noopener noreferrer'
              >
                <p className="site-menu__nav-item-text-social">Exercism</p>
                <img src={exercismLogo} alt='' />
              </a>
            </li>
            <li>
              <a
                className='site-menu__nav-item-link-social'
                href='https://www.codewars.com/users/garryunn'
                target='_blank'
                rel='noopener noreferrer'
              >
                <p className="site-menu__nav-item-text-social">Codewars</p>
                <img src={codewarsLogo} alt='' />
              </a>
            </li>
          </span>
        </span>
      </ul>
    </nav>
  );
}