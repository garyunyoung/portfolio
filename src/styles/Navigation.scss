@import './_mixins.scss';
@import './_variables.scss';

@include block('navigation') {
  width: 100%;

  background-color: $page-background-colour;

  position: fixed;
  top: 0;
  z-index: 1;

  padding-top: 24px;
  padding-bottom: 16px;

  @include sm-desktop {
    height: $navigation-height-tablet;
  }

  @include med-desktop {
    background-color: transparent;
  }

  @include state('open') {
    .navigation__nav {
      display: flex;
    }

    .navigation__mobile-menu-button::after {
      content: 'X';
    }
  }

  @include element('content') {
    max-width: calc(
      #{$max-width-desktop} + (180px * 2) + (32px * 2)
    );

    display: flex;
    flex-flow: wrap;
    align-items: center;

    margin: 0 auto;
  }

  @include element('logo-wrapper') {
    display: flex;
    flex-grow: 1;

    margin-left: $side-padding-mobile;
  }

  @include element('logo') {
    @include title-font;

    @include sm-desktop {
      padding: 12px 24px;
    }
  }

  @include element('logo-link') {
    @include call-to-action-link-button;
    @include title-font;
    padding: 0;

    background-color: $page-background-colour;

    @include sm-desktop {
      padding: 12px 24px;
    }
  }

  @include element('nav') {
    display: none;
    justify-content: flex-end;

    width: 100%;
    height: calc(100vh - #{$navigation-height-mobile});
    background-color: $page-background-colour;

    padding-top: calc(
      #{$navigation-height-mobile} + #{$spacer-large}
    );

    @include sm-desktop {
      display: block;

      height: auto;
      width: max-content;

      border: 0;
      padding-top: 0;
    }
  }

  @include element('mobile-menu-button') {
    font-size: 18px;
    line-height: 1.4;
    letter-spacing: 2px;

    border-bottom: 1px dotted transparent;
    margin: 0 $side-padding-mobile;

    &:after {
      content: 'â˜°';
      display: inline;
      margin-left: 8px;
    }

    &:focus {
      border-color: $link-underline;
    }

    &:hover {
      border-color: $link-underline;
      cursor: pointer;
    }

    @include sm-desktop {
      display: none;
    }
  }
}

@include block('navigation-menu') {
  @include element('nav-items') {
    margin-top: -$spacer-large;

    @include sm-desktop {
      display: flex;
      flex-flow: column;

      margin-top: 0;
      position: relative;
    }
  }

  @include element('nav-main-items') {
    @include sm-desktop {
      display: flex;
    }
  }

  @include element('nav-code-container') {
    display: flex;
    flex-flow: column;
    justify-content: center;
    align-items: flex-end;

    margin-top: 32px;

    img {
      width: 28px;
    }

    @include sm-desktop {
      position: absolute;
      top: 68px;
      right: 0px;

      background-color: $page-background-colour;
      padding-bottom: 8px;
      margin-top: 0;
    }
  }

  @include element('nav-item-link-social') {
    @include call-to-action-link-button;
    padding-right: 16px;

    display: flex;
    justify-content: space-between;
  }

  @include element('nav-item-text-social') {
    margin-right: 16px;
  }

  @include element('nav-item') {
    display: flex;
    justify-content: flex-end;
    align-items: center;
  }

  @include element('nav-socials') {
    display: flex;
    flex-flow: column;
    align-items: flex-end;

    @include sm-desktop {
      display: none;
    }

    @include state('visible') {
      display: flex;
    }
  }

  @include element('nav-item-link') {
    @include call-to-action-link-button;
    padding-right: 16px;

    @include modifier('socials') {
      display: none;

      @include sm-desktop {
        display: block;
        margin-bottom: 16px;

        &:before {
          display: inline-block;
          @include set-size(8px, 8px);
          content: '';
          border-top: 2px solid $black;
          border-right: 2px solid $black;
          transform: rotate(45deg);

          margin-right: 16px;
        }
      }
    }

    @include state('selected') {
      &:before {
        transform: rotate(135deg);
      }
    }
  }
}
