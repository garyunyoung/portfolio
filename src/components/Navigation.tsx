import { useState, useEffect, ReactElement } from 'react'
import { useLocation } from 'react-router-dom'

import { LinkInternal } from './partials/Links'

import github from '../assets/logos/github.svg'
import codewars from '../assets/logos/codewars.svg'
import exercism from '../assets/logos/exercism.svg'

import elementInView from '../utilities/ElementInView'

import '../styles/Navigation.scss'

interface NavigationMenuProps {
  isSocialsOpen: boolean
  toggleSocials: () => any
  closeNav: () => any
  isInView: string
}

interface LogoProps {
  isHomePage: boolean
}

interface NavItemLinkProps {
  id: string
  text: string
  href: string
  closeNav: () => any
  isInView: string
}

interface NavItemLinkSocialProps {
  text: string
  href: string
  logoSrc: string
}
export default function Navigation(): ReactElement {
  const [isOpen, setIsOpen] = useState<boolean>(false)
  const [isSocialsOpen, setIsSocialsOpen] =
    useState<boolean>(false)
  const [isInView, setIsInScrollView] =
    useState<string>('home')

  const isHomePage: boolean = useLocation().pathname === '/'

  useEffect(() => {
    if (isHomePage) {
      elementInView(setIsInScrollView)
    }
  }, [isHomePage])

  function toggleNav() {
    setIsOpen((prevState) => !prevState)
  }

  function closeNav() {
    setIsOpen(false)
  }

  function toggleSocials() {
    setIsSocialsOpen((prevState) => !prevState)
  }

  return (
    <header
      className={`navigation ${isOpen ? 'is-open' : ''}`}
    >
      <div className="navigation__content">
        <Logo isHomePage={isHomePage} />
        <button
          className="navigation__mobile-menu-button"
          onClick={() => toggleNav()}
        >
          {isOpen ? 'Close' : 'Menu'}
        </button>
        <NavigationMenu
          isSocialsOpen={isSocialsOpen}
          toggleSocials={toggleSocials}
          closeNav={closeNav}
          isInView={isInView}
        />
      </div>
    </header>
  )
}

function NavigationMenu({
  isSocialsOpen,
  toggleSocials,
  closeNav,
  isInView
}: NavigationMenuProps): ReactElement {
  return (
    <nav className="navigation-menu navigation__nav">
      <ul className="navigation-menu__nav-items">
        <span className="navigation-menu__nav-main-items">
          <NavItemLink
            text="Projects"
            href="/#projects"
            id="projects"
            closeNav={() => closeNav()}
            isInView={isInView}
          />

          <NavItemLink
            text="About"
            href="/#about"
            id="about"
            closeNav={() => closeNav()}
            isInView={isInView}
          />
        </span>
        <span className="navigation-menu__nav-code-container">
          <li onClick={() => toggleSocials()}>
            <button
              className={`
              navigation-menu__nav-item-link 
              navigation-menu__nav-item-link--socials 
              ${
                isSocialsOpen ? 'is-selected is-active' : ''
              }
            `}
            >
              Code
            </button>
          </li>
          <span
            data-testid="nav-socials"
            className={`navigation-menu__nav-socials ${
              isSocialsOpen ? 'is-visible' : ''
            }`}
          >
            <NavItemLinkSocial
              href="https://github.com/garyunyoung"
              text="Github"
              logoSrc={github}
            />

            <NavItemLinkSocial
              href="https://exercism.org/profiles/garyunyoung"
              text="Exercism"
              logoSrc={exercism}
            />

            <NavItemLinkSocial
              href="https://www.codewars.com/users/garryunn"
              text="Codewars"
              logoSrc={codewars}
            />
          </span>
        </span>
      </ul>
    </nav>
  )
}

export function Logo({
  isHomePage
}: LogoProps): ReactElement {
  return (
    <div className="navigation__logo-wrapper">
      {isHomePage ? (
        <h1 className="navigation__logo">Garyun Young</h1>
      ) : (
        <LinkInternal
          className={`navigation__logo-link ${
            isHomePage ? '' : 'is-active'
          }`}
          type="link-button"
          href="/#"
          text="Garyun Young"
        />
      )}
    </div>
  )
}
function NavItemLink({
  id,
  text,
  href,
  closeNav,
  isInView
}: NavItemLinkProps): ReactElement {
  return (
    <li className="navigation-menu__nav-item">
      <a
        className={`navigation-menu__nav-item-link ${
          isInView === id ? 'is-active' : ''
        }`}
        href={href}
        onClick={() => closeNav()}
      >
        {text}
      </a>
    </li>
  )
}

function NavItemLinkSocial({
  text,
  href,
  logoSrc
}: NavItemLinkSocialProps): ReactElement {
  return (
    <li>
      <a
        className="navigation-menu__nav-item-link-social"
        href={href}
        target="_blank"
        rel="noopener noreferrer"
      >
        <p className="navigation-menu__nav-item-text-social">
          {text}
        </p>
        <img src={logoSrc} alt="" />
      </a>
    </li>
  )
}
