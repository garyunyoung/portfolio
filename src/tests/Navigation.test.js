import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { create } from 'react-test-renderer'
import Navigation from '../components/Navigation'
import { getQueriesForElement } from '@testing-library/dom'

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    pathname: "/"
  })
}));

it('renders without crashing', () => {
  const root = document.createElement('div');
  render(<Navigation />, root);
  unmountComponentAtNode(root);
});

it('renders correct content', () => {
  const root = document.createElement('div');
  render(<Navigation />, root);

  const { getByText } = getQueriesForElement(root)

  getByText("Projects")
  getByText("About")
  getByText("Code")
});



it('matches snapshot', () => {
  const tree = create(<Navigation />).toJSON()
  expect(tree).toMatchSnapshot()
})