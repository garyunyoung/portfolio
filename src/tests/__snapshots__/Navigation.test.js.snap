// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<header
  className="navigation "
>
  <div
    className="navigation__content"
  >
    <div
      className="navigation__logo-wrapper"
    >
      <h1
        className="navigation__logo"
      >
        Garyun Young
      </h1>
    </div>
    <button
      className="navigation__mobile-menu-button"
      onClick={[Function]}
    >
      Menu
    </button>
    <nav
      className="navigation-menu navigation__nav"
    >
      <ul
        className="navigation-menu__nav-items"
      >
        <span
          className="navigation-menu__nav-main-items"
        >
          <li
            className="navigation-menu__nav-item"
          >
            <a
              className="navigation-menu__nav-item-link "
              href="/#projects"
              onClick={[Function]}
            >
              Projects
            </a>
          </li>
          <li
            className="navigation-menu__nav-item"
          >
            <a
              className="navigation-menu__nav-item-link "
              href="/#about"
              onClick={[Function]}
            >
              About
            </a>
          </li>
        </span>
        <span
          className="navigation-menu__nav-code-container"
        >
          <li
            onClick={[Function]}
          >
            <button
              className="
              navigation-menu__nav-item-link 
              navigation-menu__nav-item-link--socials 
              
            "
            >
              Code
            </button>
          </li>
          <span
            className="navigation-menu__nav-socials "
            data-testid="nav-socials"
          >
            <li>
              <a
                className="navigation-menu__nav-item-link-social"
                href="https://github.com/garyunyoung"
                rel="noopener noreferrer"
                target="_blank"
              >
                <p
                  className="navigation-menu__nav-item-text-social"
                >
                  Github
                </p>
                <img
                  alt=""
                  src="github.svg"
                />
              </a>
            </li>
            <li>
              <a
                className="navigation-menu__nav-item-link-social"
                href="https://exercism.org/profiles/garyunyoung"
                rel="noopener noreferrer"
                target="_blank"
              >
                <p
                  className="navigation-menu__nav-item-text-social"
                >
                  Exercism
                </p>
                <img
                  alt=""
                  src="exercism.svg"
                />
              </a>
            </li>
            <li>
              <a
                className="navigation-menu__nav-item-link-social"
                href="https://www.codewars.com/users/garryunn"
                rel="noopener noreferrer"
                target="_blank"
              >
                <p
                  className="navigation-menu__nav-item-text-social"
                >
                  Codewars
                </p>
                <img
                  alt=""
                  src="codewars.svg"
                />
              </a>
            </li>
          </span>
        </span>
      </ul>
    </nav>
  </div>
</header>
`;
